### PR TITLE
feat: call try_merge recursively for list field 

### LINKED ISSUE #5843: `Schema::try_merge` should be able to merge List of any data type with List of Null data type
**Describe the bug**
`Schema::try_merge` should be able to merge List(any data type) and List(Null data type)

**To Reproduce**
```rust
use serde_json::json;

let o_1 = vec![
    json!({"a": [1,2,3]}),
    json!({"a": [4,5,6]}),
];

let o_2 = vec![
    json!({"a": []}),
    json!({"a": []}),
];

let o_3 = vec![
    json!({"a": []}),
    json!({"a": [7,8,9]}),
];

let schema_1 = arrow_json::reader::infer_json_schema_from_iterator(o_1.iter().map(Ok)).unwrap();
println!("{:?}", schema_1);
// Schema { fields: [Field { name: "a", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }

let schema_2 = arrow_json::reader::infer_json_schema_from_iterator(o_2.iter().map(Ok)).unwrap();
println!("{:?}", schema_2);
// Schema { fields: [Field { name: "a", data_type: List(Field { name: "item", data_type: Null, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }

let schema_3 = arrow_json::reader::infer_json_schema_from_iterator(o_3.iter().map(Ok)).unwrap();
println!("{:?}", schema_3);
// Schema { fields: [Field { name: "a", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }

let merged = Schema::try_merge([schema_1, schema_2, schema_3]);
println!("{:?}", merged)
// Err(SchemaError("Fail to merge schema field 'a' because the from data_type = List(Field { name: \"item\", data_type: Null, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }) does not equal List(Field { name: \"item\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })"))
```

**Expected behavior**
Merge of List(<any> data type) and List(Null) should be possible
```rust
let merged = Schema::try_merge([schema_1, schema_2, schema_3]);
println!("{:?}", merged)
// Ok(Schema { fields: [Field { name: "a", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} })
```

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by mnpw**
# Which issue does this PR close?

Closes #5843.

# Rationale for this change
 
 Merging fields recursively is already performed for a field with datatype Struct. This PR extends that recursive merge to fields with List and LargeList datatypes.

# What changes are included in this PR?
- Updated `Field::try_merge` to call merge recursively if datatype is List and LargeList
- Added tests for megre for nested datatypes (like Stuct and List) with Null fields

# Are there any user-facing changes?

N/A

### CODE REVIEW

File: arrow-schema/src/field.rs [hunk 1]
```diff
@@ -483,6 +483,18 @@ impl Field {
                     ));
                 }
             },
+            DataType::List(field) => match &from.data_type {
```
Comment (alamb) [inline arrow-schema/src/field.rs:486]: I think we should also add support for `DataType::LargeList` https://docs.rs/arrow/latest/arrow/datatypes/enum.DataType.html#variant.LargeList

If you don't want to add it in this PR, I'll file a ticket to track and add it in a follow on

**Review by alamb — APPROVED** [review]
Thank you for the contribution @mnpw  -- this PR makes sense to me.

Nicely tested as well 👍 

Let me know about `DataType::LargeList`

**Comment by mnpw** [comment]
Thanks @alamb for the review. As per your suggestion I have added the change for `LargeList` too.

File: arrow-schema/src/field.rs [hunk 2]
```diff
@@ -623,6 +645,69 @@ mod test {
...
+                    "inner",
+                    DataType::Float32,
+                    true,
+                )])),
+                true,
+            ),
+            struct1
+        );
+
+        let mut list1 = Field::new(
```
Comment (alamb) [inline arrow-schema/src/field.rs:686]: Any chance we could add a test for `LargeList` too?

**Review by alamb — APPROVED** [review]
Thanks @mnpw 

I noticed that this PR is marked "breaking" but it seems to me like it isn't a breaking API change (in the sense that code that compiles before this PR will also compile after this PR)

**Comment by mnpw** [comment]
@alamb 

- Added test for `LargeList`
- Updated commit to reflect the non-breaking nature of change. 

I added `breaking` to highlight the different outcome `Field::try_merge` after my changes. It seems more fit to only add `breaking` if there are user-facing API changes as you suggest.

Lmk if this looks good

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again @mnpw

**Pull request merged**